[FLINK-22748][connector-kafka] Allow dynamic target topic selection in SQL Kafka sinks

SteNicholas opened the pull request:
## What is the purpose of the change

*We should allow to write to different Kafka topics based on some column value in SQL. The existing implementation can be easily adapted for that. The "target topic" would be an additional persisted metadata column in SQL terms. All one need to do is to adapt `DynamicKafkaSerializationSchema` and `KafkaDynamicSink`. We should guard this dynamic behavior via a metadata column and make the topic option optional in this case.*

## Brief change log

  - *`KafkaDynamicSink` adds the `topic` writable metadata column and make the `topic` option optional for `KafkaDynamicTableFactory`.*
  - *`DynamicKafkaSerializationSchema` serializes the row data to `ProducerRecord` via `topic` metadata column when the 'topic' table option is null.*

## Verifying this change

  - *`KafkaTableITCase` adds the `testKafkaSinkWithMetadataIncludeTopicOption` and `testKafkaSinkWithMetadataExcludeTopicOption` test methods to verify whether sink to Kafka through `topic` metadata column when the 'topic' table option is null.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

SteNicholas commented:
@twalthr , could you please help to review this pull request?

SteNicholas commented:
@flinkbot run azure

Airblader commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/DynamicKafkaSerializationSchema.java:
```diff
@@ -166,6 +167,15 @@ public void setPartitions(int[] partitions) {
     @Override
     public String getTargetTopic(RowData element) {
+        // no explicit topic defined in table options, use the writable metadata column topic
+        topic =
```
What happens / should happen if the topic returned by metadata differs between different elements? We're now essentially reading the metadata of whichever element comes first and then ignore it forever, right? That isn't the behavior I would expect given the documentation.

SteNicholas replied:
If the topic returned by metadata differs between different elements, the corresponding record could produced to different topic. I don't understand where isn't the behavior I would expect given the documentation.

Airblader replied:
When `getTargetTopic(firstElement)` is called and `topic == null`, this will assign the value from the metadata to the topic. Now if `getTargetTopic(secondElement)` is called, `topic` will no longer be `null` and thus the metadata will not be read from the second element, so no matter what the metadata of this second element is, `topic` will be that of the first element still. Is that correct?

SteNicholas replied:
@Airblader , yes, your point is correct. I would update the commit for this.

Airblader commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactory.java:
```diff
@@ -187,6 +188,9 @@ public DynamicTableSource createDynamicTableSource(Context context) {
     @Override
     public DynamicTableSink createDynamicTableSink(Context context) {
+        // validate the table option 'topic' and 'topic-pattern'
+        validateSinkTopic(Configuration.fromMap(context.getCatalogTable().getOptions()));
```
Use `helper.getOptions()` (already stored in a variable `tableOptions` below) instead of building another `Configuration` here? Is there a reason to have this on the very top and not e.g. where `validateTableSinkOptions` is called?

Airblader commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaOptions.java:
```diff
@@ -361,18 +360,25 @@ public static void validateSourceTopic(ReadableConfig tableOptions) {
-                throw new ValidationException(
-                        String.format(errorMessageTemp, "'topic'", tableOptions.get(TOPIC)));
+                                "Flink Kafka sink doesn't support empty topic, but got %s: %s.",
+                                "'topic'", tableOptions.get(TOPIC)));
```
nit: it isn't really necessary to call `tableOptions.get(TOPIC)` here again since this will be called exactly when the list is empty.

Airblader commented on flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java:
```diff
@@ -300,6 +306,240 @@ public void testKafkaTableWithMultipleTopics() throws Exception {
+
+        // ------------- cleanup -------------------
+
+        deleteTestTopic(topic);
```
We should make sure this will always be called as soon as `createTestTopic` has been called. Right now there's a few things that could throw exceptions in between and would prevent this from being called.

SteNicholas replied:
@Airblader , the unit tests added could ensure the `deleteTestTopic(topic)` called after `createTestTopic`.

Airblader replied:
This was marked as resolved, but I don't see any changes here. If there's a failure or a wrong assertion, the cleanup won't be called.

SteNicholas replied:
@Airblader , I replied to the comment above and no any things could prevent `deleteTestTopic` being called in the added test cases.

Airblader replied:
Almost everything here can prevent `deleteTestTopic` from being called. `tEnv.executeSql` can throw an exception, any assertion which fails would prevent it from being called (since failing assertions throw exceptions). Basically this all needs to be wrapped into try-finally:

```
try {
  createTestTopic();

 // …
} finally {
  deleteTestTopic();
}
```

But I'm also just noticing that all the existing tests in this class have the same issue. :-(

SteNicholas replied:
@Airblader , does this problem need to be fixed in this pull request?

Airblader replied:
I think it'd be good if we can at least fix it for the new tests you added, but we don't need to change the existing tests as well.

SteNicholas replied:
@Airblader , I would like to wrap the new tests added in the pull request.

SteNicholas commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java:
```diff
@@ -216,7 +216,7 @@
 
     /** The name of the default topic this producer is writing data to. */
-    protected final String defaultTopicId;
+    protected String defaultTopicId;
```
@Airblader , `defaultTopicId` would be initialised in the `invoke` method for the case that null value for `KafkaOption#TOPIC`.

SteNicholas replied:
@twalthr , this is a valid change because `defaultTopicId` would be updated in the `invoke` method for the case that null value for `KafkaOption#TOPIC`.

SteNicholas commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/DynamicKafkaSerializationSchema.java:
```diff
@@ -166,6 +167,15 @@ public void setPartitions(int[] partitions) {
     public String getTargetTopic(RowData element) {
+        // no explicit topic defined in table options, use the writable metadata column topic
+        topic =
+                StringUtils.isNullOrWhitespaceOnly(topic)
```
@Airblader , the `KafkaOptions#validateSinkTopic` could ensure that whitespace-only is mapped to null. And the `KafkaOptions#TOPIC` couldn't be whitespace-only.

SteNicholas commented:
@Airblader , thanks for your detailed review. I have adressed your comments and left my comments. Please help to review again.

SteNicholas commented:
@Airblader , thanks for your detailed review. I have adressed your comments and left my comments. Please help to review again.

twalthr reviewed (commented):
Thanks @SteNicholas. I left a couple of comments. Sorry for being so picky but the Kafka connector is the most important connector in Flink, so we need to make this bullet proof. I would suggest to have slightly different semantics for sinks. The `topic` option could be the list of allowed values for the metadata column. If `topic` option is not defined any value is supported in the metadata column. What do you think?

SteNicholas commented:
@twalthr , If topic option is not defined, what's the behavior when null value or emty string for metadata column?

SteNicholas commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactory.java:
```diff
@@ -224,7 +224,7 @@ public DynamicTableSink createDynamicTableSink(Context context) {
                 valueProjection,
                 keyPrefix,
-                tableOptions.get(TOPIC).get(0),
+                tableOptions.getOptional(TOPIC).isPresent() ? tableOptions.get(TOPIC).get(0) : null,
```
@twalthr , this list couldn't be empty because this case is checked in `KafkaOptions#validateSinkTopic` method.

AHeise replied:
Here at this place, you could use `unspecified!` instead of null. Then you don't need to change anything in the `KafkaProducer`. If you use a special character like `!`, this will fail hard if the user row does not specify the column.

SteNicholas replied:
@AHeise , I agree with this approach but I'm confused about why not this uses null? I don't understand the difference between `unspecified!` and null.

SteNicholas commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaOptions.java:
```diff
@@ -144,7 +144,7 @@ private KafkaOptions() {}
                     .withDescription(
                             "Topic names from which the table is read. Either 'topic' or 'topic-pattern' must be set for source. "
-                                    + "Option 'topic' is required for sink.");
+                                    + "Option 'topic' is optional for sink through specifying the 'topic' metadata column.");
```
@twalthr , this is originally different from the doc. Please check again.

twalthr replied:
Yes, because many people don't now how to generate docs from options. Currently, it still involves some manual work but we will fix this issue soon.

Airblader replied:
This comment has been marked as resolved, but the documentation still seems to differ from the description in the ConfigOption.

SteNicholas replied:
@Airblader, IMO, currently the description is different from the options on the document. My understanding is that @twalthr would unify the description of the options and the options on the document. Therefore I didn't update this.

Airblader replied:
I think he meant we will fix the issue of this being manual work, but if we touch it now we should still make sure to align it.

SteNicholas commented:
> Thanks @SteNicholas. I left a couple of comments. Sorry for being so picky but the Kafka connector is the most important connector in Flink, so we need to make this bullet proof. I would suggest to have slightly different semantics for sinks. The `topic` option could be the list of allowed values for the metadata column. If `topic` option is not defined any value is supported in the metadata column. What do you think?

@twalthr , thanks for your detailed review and I have addressed your comments for the update. IMO, if `topic` table option isn't defined, the value of the `topic` writable metadata column could accept any value and if the value is null or empty, this could be corrected by user self.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java:
```diff
@@ -216,7 +216,7 @@
 
     /** The name of the default topic this producer is writing data to. */
-    protected final String defaultTopicId;
+    @Nullable protected String defaultTopicId;
```
I don't understand why we need to make default topic nullable. We could just set "unspecified" on table API if there is no other value. Potentially add some special character that kafka doesn't support to make sure writing fails.
As is, we change quite a bit of API and semantics and introduce new bugs (there is no guard against `defaultTopicId = null` and `getTargetTopic = null`).
Another idea would be to explicitly support default topic option with said default value.

SteNicholas replied:
@AHeise , if the user doesn't define the `TOPIC` table option, the value of `defaultTopicId` in the constructor is null and the value of `defaultTopicId` is depended on the `getTargetTopic`. the return value of `getTargetTopic` is depended on the value of `TOPIC` metadata column, and @twalthr mentioned that If `TOPIC` option is not defined any value is supported in the metadata column. For @AHeise 's opinion, if `TOPIC` option is not defined and the value of `TOPIC` metadata column is null, the `defaultTopicId` should be a default value like `unspecified`, right?

SteNicholas commented:
@Airblader @AHeise , I have addressed the comments from @twalthr , could you please take a look at the changes?

AHeise reviewed (commented):
I have added a possible change further down that allows you to leave `KafkaProducer` untouched.

SteNicholas commented:
> I have added a possible change further down that allows you to leave `KafkaProducer` untouched.

@AHeise , thanks for the detailed advice and I have used `unspecified!` as default topic name. Please help to review again.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java:
```diff
@@ -860,7 +861,7 @@ public void invoke(
             if (targetTopic == null) {
                 targetTopic = defaultTopicId;
-            } else if (defaultTopicId == null) {
+            } else if (TOPIC_UNSPECIFIED.equals(defaultTopicId)) {
```
Revert.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java:
```diff
@@ -900,7 +901,7 @@ public void invoke(
                 if (targetTopic == null) {
                     targetTopic = defaultTopicId;
-                } else if (defaultTopicId == null) {
+                } else if (TOPIC_UNSPECIFIED.equals(defaultTopicId)) {
```
Revert.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/DynamicKafkaSerializationSchema.java:
```diff
@@ -166,7 +168,7 @@ public void setPartitions(int[] partitions) {
     public String getTargetTopic(RowData element) {
-        return topic == null
+        return TOPIC_UNSPECIFIED.equals(topic)
                 ? readMetadata(element, KafkaDynamicSink.WritableMetadata.TOPIC)
```
I just saw that `readMetadata` can return `null` which it shouldn't (and it's not annotated!). 
What is the expected target topic then? I guess `TOPIC[0]`?
But what happens if the user didn't set `TOPIC`? Can you check the schema if the column is nullable and fail if `TOPIC` is empty/unset?

twalthr replied:
This comment is still not addressed. When neither option nor metadata column are defined, `readMetadata` will return null and cause Kafka to throw an exception. We can be smarter here and fail already during planning. In `KafkaDynamicSink#getSinkRuntimeProvider` we have all information to create a helpful exception if the metadata column is missing.

AHeise reviewed (changes requested):
We are nearly there. Let me try to rephrase once more what I had in mind:
- `defaultTopicId` in `FlinkKafkaProducer` is used when `KafkaSerializationSchema#getTargetTopic` returns `null`
- So a user can override the topic with `getTargetTopic` or just use the `defaultTopicId` (this is especially true with `KeyedSerializationSchema`)
- Now table API is just another "user". However, we know that `DynamicKafkaSerializationSchema#getTargetTopic` is never returning `null`. So whatever we use as `defaultTopicId`, it doesn't matter.
- Adding `TOPIC_UNSPECIFIED` is a good idea but it should be seen as an implementation of the Table API user.
- No special handling of that value is needed in `FlinkKafkaProducer`; it's always overridden.
- To detect any bugs, `TOPIC_UNSPECIFIED` uses `!`, such that in the case that no topic is associated with a `Row`, we fail hard (invalid topic name). You could also add an additional `checkState` in `DynamicKafkaSerializationSchema#getTargetTopic/readMetadata`

I additionally saw that the user actually can make `DynamicKafkaSerializationSchema#getTargetTopic` return null; which shouldn't happen imho. Please enforce that a user always provides a topic iff `TOPIC` is unset.

SteNicholas commented:
@AHeise , thanks for detailed rephrase and I hold a different point of view:

1.  The `defaultTopicId` could be null in the constructor of `KafkaSerializationSchema` for the case that user doesn't define the `TOPIC` table option.
2. The return value of `DynamicKafkaSerializationSchema#getTargetTopic` is never null after adding the check whether the `TOPIC` metadata column is null or empty. But this check is a very expensive operation that is executed for every record in the hot path, therefore in terms of performance this check could be unnecessary and can rely on the user to specify it correctly.

I have added the check whether the `TOPIC` metadata column is null or empty in `DynamicKafkaSerializationSchema#getTargetTopic/readMetadata` but IMO this check could be removed. @AHeise what do you think about?

AHeise commented:
> @AHeise , thanks for detailed rephrase and I hold a different point of view:
> 
>     1. The `defaultTopicId` could be null in the constructor of `KafkaSerializationSchema` for the case that user doesn't define the `TOPIC` table option, in which case that `defaultTopicId` is unnecessary regarded as `TOPIC_UNSPECIFIED `.

It's a bad habit to weaken invariants to add new features, especially if they are on different abstraction levels. Don't forget that `FlinkKafkaProducer` is very commonly used in DataStream applications. Why make it harder to detect bugs in DataStream applications, just so that Table API has a new feature?

Instead of weakening invariants, please always consider if there is a way to express your intention without modification of low layers. From the Table API's perspective, the actual value of `defaultTopicId` doesn't matter, so there is a large degree of freedom that you should use.

If you insist on `defaultTopicId` being nullable, then you need to guard `FlinkKafkaProducer` against all cases of `defaultTopicId == null && getTargetTopic() == null`. There are at least two code-paths that can lead to the violated invariant `topicId == null`, so you need to provide good error-messages that the user can quickly see which code-path lead to the violation. Further, I expect test coverage of all cases then before I can approve. Imho that's wasted effort.

>     2. The return value of `DynamicKafkaSerializationSchema#getTargetTopic` is never null after adding the check whether the `TOPIC` metadata column is null or empty. But this check is a very expensive operation that is executed for every record in the hot path, therefore in terms of performance this check could be unnecessary and can rely on the user to specify it correctly.

I was hoping that we can do it statically: when setting up the Flink job, don't we have a way to detect if a column is nullable or not? You could simply reject an input where the column could potentially contain nulls, no? 

If that is impossible, we could simply reduce the check to `isNull` (that's rather cheap), I don't see why we need to check for empty values. We are also not verifying if the string is a valid topic name at this point, so just relay that to Kafka.

SteNicholas commented:
@AHeise , thanks for good suggestion. It's my mistake to weaken invariants for new features. I have addressed the comments you mentioned and the added tests have already covered all cases. Please check again for the update.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java:
```diff
@@ -856,6 +856,8 @@ public void invoke(
             if (targetTopic == null) {
                 targetTopic = defaultTopicId;
+            } else {
+                defaultTopicId = targetTopic;
```
Revert

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java:
```diff
@@ -894,6 +896,8 @@ public void invoke(
                 if (targetTopic == null) {
                     targetTopic = defaultTopicId;
+                } else {
+                    defaultTopicId = targetTopic;
```
Revert

AHeise reviewed (commented):
There are still some changes in FlinkKafkaProducer that you should revert. Please also squash the commits. Then I'd do a final pass.

SteNicholas commented:
> There are still some changes in FlinkKafkaProducer that you should revert. Please also squash the commits. Then I'd do a final pass.

Yes, I have reverted the `FlinkKafkaProducer` and squashed the commits. Please review the squash.

Airblader reviewed (changes requested):
Just following up on a couple of comments which don't seem to be addressed as far as I can tell. I also think @twalthr is going to want to give this a final review.

AHeise reviewed (approved):
Thank you very much for your contribution. LGTM (from the DataStream's perspective).

SteNicholas commented:
> Just following up on a couple of comments which don't seem to be addressed as far as I can tell. I also think @twalthr is going to want to give this a final review.

@Airblader , the two comments has already addressed and I have squashed the commits. Please check again.

twalthr reviewed (changes requested):
Sorry @SteNicholas but I had to add a couple of concerns. We need to improve the quality of this contribution further.

SteNicholas commented:
> Sorry @SteNicholas but I had to add a couple of concerns. We need to improve the quality of this contribution further.

@twalthr , sorry for the later update of this pull request. I have addressed the comments mentioned above. Please help to review againn.

benoitparis commented:
Sorry for chiming in late in the conversation.

By curiosity @twalthr when you say 'People would like to emit to a list of topics', is it to emit a message to topics in a round-robin way, or is it to duplicate it to several topics?

klam-shop commented:
Hey saw this PR is closed, is any work being done still? Interested in this feature :)

klam-shop commented:
👋 I'm looking to pick up this work. I can move it to https://github.com/apache/flink-connector-kafka since the Kafka Connector code has been moved there.

If I understand correctly there are a few outstanding things to resolve before this feature can be committed:

1. Address the [comment here](https://github.com/apache/flink/pull/16142#pullrequestreview-783262121) to support the 'whitelist' behaviour @twalthr described.
2. Clean up tests in `flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java`
3. There may be some other things that come up as we move it to the `flink-connector-kafka` repo, as this PR is quite stale.

Does the above seem correct?

klam-shop commented:
Hi I've taken some time to build upon [Nicholas Jiang](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=nicholasjiang) 's PR and port it to the new kafka connector repo:

https://github.com/apache/flink-connector-kafka/pull/109

Would appreciate feedback on this approach, let me know.
